Added slew rate limiter to AC_PID

tridge opened the pull request:
This is based on the slew limiter that @priseborough did for fixed wing, applied to AC_PID. This is part of the effort to move fixed wing to using AC_PID, but it is also directly useful in copters.
It adds two new parameters to all AC_PID objects, an SMAX and STAU. If SMAX is set to a non-zero value then it will slew limit the P+D value, preventing oscillation due to too high gains. The modifier (between 0.1 and 1) is logged in the PID logs as Dmod.

Demonstration in RF9 for copter here:
https://youtu.be/8qO1FgZhH04
and for heli here:
https://www.youtube.com/watch?v=eYkjD-rDOOg
Needs testing on some real vehicles

tridge commented:
@priseborough and @lthall I'd like to get your input on whether we really need the time constant parameter for decay of the filter exposed to users. Could we use 1s for all AC_PID users?
Also, the implementation uses a fixed 10Hz filter internally. Should we set this equal to the FLTE parameter, which is the filter we apply on the error term in AC_PID? Or maybe FLTT, the target filter?
I've also found a fairly easy was to see what slew rates we are producing in our existing logs, so we can start to work out sensible defaults. The RATE message is at the full loop rate, and is on the same scale as this filter, so we can just do graphs like this:
![image](https://user-images.githubusercontent.com/831867/94194112-94ca9080-fef4-11ea-89cb-7860fc4fcd48.png)
that is a log from a tiny copter (a LuminousBee 3" quad doing fast flying of a Lissojous curve in a Vicon room).
Compare that to this log of a 12" quad flying a fairly tame auto mission:
![image](https://user-images.githubusercontent.com/831867/94194676-5d101880-fef5-11ea-8773-485c10025522.png)
and here is a 28" octaquad carrying a 3kg load then dropping it:
![image](https://user-images.githubusercontent.com/831867/94194898-aa8c8580-fef5-11ea-8e90-d4e3cd3126d7.png)
the variation in the slew rates is not really that large.
On the vertical axis we see much lower slew rates. Here is the LuminousBee 3":
![image](https://user-images.githubusercontent.com/831867/94195168-035c1e00-fef6-11ea-831c-e2644dba6c17.png)
and here is the 28" octaquad:
![image](https://user-images.githubusercontent.com/831867/94195281-2555a080-fef6-11ea-8a0a-f3e2179e4a0f.png)
So it looks like we could set the vertical accel default SMAX quite low, maybe 5 to 10? The SMAX default for roll and pitch would need to be something like 50 to not impact existing vehicles.

tridge commented on AntennaTracker/Parameters.cpp:
```diff
@@ -355,6 +355,21 @@ const AP_Param::Info Tracker::var_info[] = {
+    // @Increment: 0.5
+    // @User: Advanced
+
+    // @Param: PITCH2SRV_STAU
```
change to STC

rmackay9 commented on AntennaTracker/Parameters.cpp:
```diff
@@ -416,6 +431,21 @@ const AP_Param::Info Tracker::var_info[] = {
     // @User: Standard
+
+    // @Param: YAW2SRV_SMAX
+    // @DisplayName: Yaw slew rate limit
```
Is it "Yaw Rate Slew Limit" perhaps?

bnsgeyer replied:
@rmackay9 I don't know exactly how this works but if it is limiting the rate of the command out to the actuator, then it is a slew rate limit.  It is not limiting yaw rate.

rmackay9 commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Heli.cpp:
```diff
@@ -144,6 +159,21 @@ const AP_Param::GroupInfo AC_AttitudeControl_Heli::var_info[] = {
     // @User: Standard
+
+    // @Param: RAT_PIT_SMAX
+    // @DisplayName: Pitch slew rate limit
```
Let's change the display name to, "Pitch Rate Slew Limit" I think..

bnsgeyer replied:
@rmackay9 my comment is the same here as it is for yaw slew rate limit

bnsgeyer replied:
@rmackay9  this is limiting the rate of the command out to the actuator.  it is a slew rate limit not a pitch rate limit.  I deleted my comment for antenna tracker because that might command rate but I would still think that it should be called slew rate limiter even in that case.

rmackay9 replied:
OK, as long as some thought has been put into it..

tridge commented:
I did some test flights today on a small quad (a Holybro 250)
![image](https://user-images.githubusercontent.com/831867/95008248-de894880-0663-11eb-95a0-9aa593c6345d.png)
I've put two flight logs here:
http://uav.tridgell.net/HolyQuad2/2020-10-04/
I was surprised to find that it flew fine with a RLL and PIT SMAX down to 5. I also tested deliberately pushing up the D gains to huge levels and the showed that the SMAX did stop the oscillation. The logs contain lots of changes to gains and SMAX levels.
At the end of the first flight I turn off the SMAX for RLL and it immediately went out of control and crashed.

lthall commented:
I have finally had some clear thinking time to go over this.

The Good:
- I like the concept and I have no doubt that it works as advertised. 
- This could be useful to increase safety for people during their initial fights and when payloads cause a controller resonance.
- This may also help things like ground resonances where the system response has fundamentally changed and I currently have not way of addressing these issues.

My Concerns:
- Extra parameters. I agree with Tridge that we can probably drop the recovery time constant and hard code it to 1 second.
- Frequency response. It is currently hard coded to 10 Hz. It is probably more appropriate to use the FLTD and pass that into the filter.
- Behaviour during normal flight........

Frequency response: Currently the slew rate is filtered at 10 Hz. This is high for many aircraft but no so much for multirotors. Even large multirotors can see D term oscillation significantly higher than this. One of the main failure modes for small multirotors is high frequency oscillations causing motors or ESC's to get hot and burn out. If we use the FLTD number we can set this filter to match the expected upper frequency range of the controller.

Normal flight: This is a little tricky because we don't know what we don't know. 
- I would like to see Andy fly this on his small quad to see what it does in normal left-turn-only flights on an aircraft tuned stupidly fast (one with decent logging). 
- We would need to do is ensure Autotune turned this feature off.
- Noise... I hope that Andy's flight will give us an idea if noise will trigger this.

andyp1per commented:
So did six test flights today:

- The first two were on stock firmware with the same tune for comparison purposes. One of these I managed to get a battery failsafe in the middle of a roll which I also caught on camera. I can confirm that we successfully failsafe even when rolling ;)
- SMAX = 10 - this was unflyable, massive overshoot in PosHold and clear overshoot at the edges in stabilise. I could have easily crashed the copter without much trouble so did not do much other than verify it was no good.
- SMAX = 100 - LOS stabilize test looked ok, no obvious instability and crisp response, but acro felt horrible, really sloshy and not locked in at all. Did not do much more than fly around.
- SMAX = 150 - this felt and flew ok. Could roll and flip ok
- SMAX = 120 - this also didn't feel too bad, but not quite as locked in perhaps.

Will post logs later

tridge commented:
>     * SMAX = 100 - LOS stabilize test looked ok, no obvious instability and crisp response, but acro felt horrible, really sloshy and not locked in at all. Did not do much more than fly around.

I think there is a problem in your testing. For the two logs where you had SMAX=100, the Dmod in the PIDR and PIDP logs was always 1.0. That means it had zero impact on your flight.
![image](https://user-images.githubusercontent.com/831867/95694101-dccf0e80-0c7b-11eb-9989-abaf30b28dcf.png)
compare this to the flight where you had SMAX=10
![image](https://user-images.githubusercontent.com/831867/95694111-f7a18300-0c7b-11eb-800f-0718b8057236.png)
Looking at your actual rates in the SMAX=100 flight we see this:
![image](https://user-images.githubusercontent.com/831867/95694153-1ef85000-0c7c-11eb-9811-896e75d630b9.png)
so we would expect to start to see an impact visible to the pilot with SMAX=50, and values down to 30 or so should fly fine. It is not surprising that 10 is far too soft.

tridge commented:
> SMAX = 120 - this also didn't feel too bad, but not quite as locked in perhaps.

again, the filter didn't do anything except for a moment after you hit the ground on landing. Pilot experience is difficult to gauge :-)

tridge commented:
I've updated the PR to remove the STAU params, but left it available as an option for a parent class holding the AC_PID to have the param (so we can have it for plane if @priseborough  wants it retained)
Also fixed fabsf() bug in the filter that @lthall and I found

tridge commented:
>     * Frequency response. It is currently hard coded to 10 Hz. It is probably more appropriate to use the FLTD and pass that into the filter.

as discussed, with the fabsf() fix I think we should leave this as 10. 

> We would need to do is ensure Autotune turned this feature off.

are you sure about this? If someone does have it set to a low value (say 20), then don't we want autotune to take account of that?
